Add bluelink connector

kupad opened the pull request:
## External Context

https://solidarityco.atlassian.net/browse/BLUE-2340

## Description

Adds in a Bluelink connector to parsons

kupad committed:
add bluelink connector and tests

kupad committed:
add document

kupad committed:
add bluelink to doc index

kupad committed:
update API_URL

kupad committed:
remove Flags

kupad committed:
obey flake8

kupad committed:
Merge branch 'master' into add-bluelink-connector

kupad commented on parsons/bluelink/__init__.py:
```diff
@@ -0,0 +1,7 @@
+
+__all__ = [
+    'Bluelink',
+    'Person', 'Email', 'Address', 'Phone', 'Identifier', 'Tag', 'Score'
```
I'm not sure if it's a good idea to put these here, or if they are here if they should have names like BluelinkPerson instead of Person. Will spend a bit of time looking at this.

kupad replied:
Okay, I think the way I have it here is good.

Parsons has an (I think optional) convention where you can export out your package's classes from parsons' own package.
see: https://github.com/BluelinkData/parsons/blob/master/parsons/__init__.py

 It would allow you to do this:
`from parsons import Bluelink, Person ...`

instead of:
`from parsons.bluelink import Bluelink, Person`

Doing that loses some of the benefits of namespacing. A name like Person is quite likely to clash with another class in the future. So then all our classes would need to be renamed (BluelinkPerson, BluelinkEmail....).

I don't think we need to export our classes from the parsons package (it's not suggested in the 'build a connector' documentation) so I'm going to stick with these names.

kupad committed:
Update bluelink.rst

pycharm and github have different ideas on how to do code blocks in an rst. 

github wins in this case, since seeing the documentation correctly there is so important

kupad committed:
Update bluelink.rst

kupad committed:
Update bluelink.rst

lalpert commented on parsons/bluelink/bluelink.py:
```diff
@@ -0,0 +1,86 @@
+        self.user = check_env.check('BLUELINK_WEBHOOK_USER', user)
+        self.password = check_env.check('BLUELINK_WEBHOOK_PASSWORD', password)
+        print(self.user)
+        print(self.password)
```
remove these

kupad replied:
oy

lalpert commented on parsons/bluelink/bluelink.py:
```diff
@@ -0,0 +1,86 @@
+
+    def upsert_person(self, source, person=None):
+        """
+        Upsert A person object into Bluelink.
```
minor typo -- "A" should be lowercase

lalpert commented on parsons/bluelink/bluelink.py:
```diff
@@ -0,0 +1,86 @@
+    def upsert_person(self, source, person=None):
+        """
+        Upsert A person object into Bluelink.
+        For an update:
```
Should this say "For an upsert" to be consistent?

kupad replied:
I'll try to make it more clear that I'm trying to indicate what the conditions are for an update instead of an insert.

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+        self.employer = employer
+        self.employer_address = employer_address
+        self.occupation = occupation
+        self.birthdate = birthdate  # TODO: verify is ISO
```
Probably makes sense to remove these TODOs before making the final PR?

kupad replied:
Okay, I removed them. 

There aren't many TODOs in the code base, just a few. Personally, I think it would have been okay to leave em there, but I also understand why we might not have wanted to.

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+            additional custom data. must be json serializable
+    """
+    def __init__(self, identifiers, given_name=None, family_name=None, phones=None, emails=None,
+                 addresses=None, tags=None, flags=None, employer=None, employer_address=None,
```
remove `flags` here

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+    Used for to upserting via Bluelink connector.
+    See: https://bluelinkdata.github.io/docs/BluelinkApiGuide#person-object
+
+    `Args:`
```
Minor, but there are a few places the order of args here doesn't match with the order in the function

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+        return self.__json__()
+
+    @staticmethod
+    def from_table(tbl, dict_to_person):
```
Optional: Not sure I see the benefit of having this as a function on Person. Maybe we should just inline this line where the function is called.

kupad replied:
I know it's just one list comprehension, but I like giving it a name for clarity. Also, I like that it's something that can be unit tested this way

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+        source: str
+            External system to which this ID belongs, e.g., “VAN:myCampaign”.
+            Bluelink has standardized strings for source. Using these will allow
+            Lightrail to correctly understand the external IDs you add. source
```
We're not using the name "Lightrail" externally anymore. Can just change to "Bluelink"

kupad replied:
Oh! I didn't know that.

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+        address_lines: list[str]
+            A list of street address lines.
+        city: str
+            City of other locality
```
of -> or

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+        state: str
+            State in ISO 3166-2
+        postal_code: str
+            zip or other postal code
```
Probably capitalize Zip for consistency

lalpert commented on parsons/bluelink/person.py:
```diff
@@ -0,0 +1,242 @@
+        family_name: str
+            Last name / family name
+        emails: list[Email]
+            A list of Email objects representing email addresses.
```
Can you update all the args descriptions to be consistent with punctuation? Either end all in `.` or none

kupad replied:
I'll make it a task to read up on technical documentation style. I'm not sure what the best practice to do in certain situations, like when ending a sentence with a quote character, where the quotes are being used to indicate a literal. Other than those weird situations, I changed it so that every sentence ends in a period.

lalpert reviewed (commented):
Looks great! Added a bunch of nitpicks on spelling and grammar since we're going to publish this externally. (Sorry if I was premature in reviewing for that)

kupad committed:
many minor improvements (and 1 bigger one)

1- removed debugging prings
2- spelling and grammar
3- don't pass in [] as a default in python!

kupad committed:
Merge branch 'add-bluelink-connector' of github.com:BluelinkData/parsons into add-bluelink-connector
